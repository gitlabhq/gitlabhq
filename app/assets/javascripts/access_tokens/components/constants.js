import { __, s__ } from '~/locale';

export const EVENT_ERROR = 'ajax:error';
export const EVENT_SUCCESS = 'ajax:success';
export const FORM_SELECTOR = '#js-new-access-token-form';

export const INITIAL_PAGE = 1;
export const PAGE_SIZE = 100;

const BASE_FIELDS = [
  {
    key: 'name',
    label: __('Token name'),
    isRowHeader: true,
    sortable: true,
  },
  {
    formatter(description) {
      return description ?? '-';
    },
    key: 'description',
    label: __('Description'),
    sortable: true,
  },
  {
    formatter(scopes) {
      return scopes?.length ? scopes.join(', ') : __('no scopes selected');
    },
    key: 'scopes',
    label: __('Scopes'),
    sortable: true,
  },
  {
    key: 'createdAt',
    label: s__('AccessTokens|Created'),
    sortable: true,
  },
  {
    key: 'lastUsedAt',
    label: __('Last Used'),
    sortable: true,
  },
];

const ROLE_FIELD = {
  key: 'role',
  label: __('Role'),
  sortable: true,
};

export const FIELDS = [
  ...BASE_FIELDS,
  {
    formatter(ips) {
      return ips?.length ? ips?.join(', ') : '-';
    },
    key: 'lastUsedIps',
    label: __('Last Used IPs'),
    sortable: false,
    tdClass: 'gl-break-all gl-min-w-20',
  },
  {
    key: 'expiresAt',
    label: __('Expires'),
    sortable: true,
  },
  ROLE_FIELD,
  {
    key: 'action',
    label: __('Action'),
    tdClass: '!gl-py-3',
  },
];

export const INACTIVE_TOKENS_TABLE_FIELDS = [
  ...BASE_FIELDS,
  {
    key: 'expiresAt',
    label: __('Expired'),
    sortable: true,
  },
  ROLE_FIELD,
];
