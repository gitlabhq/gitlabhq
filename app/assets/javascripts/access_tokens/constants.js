import { GlFilteredSearchToken } from '@gitlab/ui';
import { s__, __ } from '~/locale';
import {
  OPERATORS_AFTER_BEFORE,
  OPERATORS_IS,
} from '~/vue_shared/components/filtered_search_bar/constants';
import DateToken from '~/vue_shared/components/filtered_search_bar/tokens/date_token.vue';

// Token types
export const FEED_TOKEN = 'feedToken';
export const INCOMING_EMAIL_TOKEN = 'incomingEmailToken';
export const STATIC_OBJECT_TOKEN = 'staticObjectToken';

export const DEFAULT_FILTER = [
  {
    type: 'state',
    value: {
      data: 'active',
      operator: '=',
    },
  },
];

export const FILTER_OPTIONS = [
  {
    icon: 'status',
    title: s__('AccessTokens|State'),
    type: 'state',
    token: GlFilteredSearchToken,
    operators: OPERATORS_IS,
    unique: true,
    options: [
      { value: 'active', title: s__('AccessTokens|Active') },
      { value: 'inactive', title: s__('AccessTokens|Inactive') },
    ],
  },
  {
    icon: 'remove',
    title: s__('AccessTokens|Revoked'),
    type: 'revoked',
    token: GlFilteredSearchToken,
    operators: OPERATORS_IS,
    unique: true,
    options: [
      { value: 'true', title: __('Yes') },
      { value: 'false', title: __('No') },
    ],
  },
  {
    icon: 'history',
    title: __('Created date'),
    type: 'created',
    token: DateToken,
    operators: OPERATORS_AFTER_BEFORE,
    unique: true,
  },
  {
    icon: 'history',
    title: __('Expiration date'),
    type: 'expires',
    token: DateToken,
    operators: OPERATORS_AFTER_BEFORE,
    unique: true,
  },
  {
    icon: 'history',
    title: __('Last used date'),
    type: 'last_used',
    token: DateToken,
    operators: OPERATORS_AFTER_BEFORE,
    unique: true,
  },
];

export const FILTER_OPTIONS_CREDENTIALS_INVENTORY = [
  {
    icon: 'key',
    title: s__('CredentialsInventory|Type'),
    type: 'filter',
    token: GlFilteredSearchToken,
    operators: OPERATORS_IS,
    unique: true,
    options: [
      {
        value: 'personal_access_tokens',
        title: s__('CredentialsInventory|Personal access tokens'),
      },
      { value: 'ssh_keys', title: s__('CredentialsInventory|SSH keys') },
      {
        value: 'resource_access_tokens',
        title: s__('CredentialsInventory|Project and group access tokens'),
      },
      { value: 'gpg_keys', title: s__('CredentialsInventory|GPG keys') },
    ],
  },
  {
    icon: 'user',
    title: s__('CredentialsInventory|Owner type'),
    type: 'owner_type',
    token: GlFilteredSearchToken,
    operators: OPERATORS_IS,
    unique: true,
    options: [
      { value: 'human', title: __('Human') },
      { value: 'service_account', title: __('Service account') },
    ],
  },
  ...FILTER_OPTIONS,
];

export const DEFAULT_SORT = { value: 'expires', isAsc: true };

export const SORT_OPTIONS = [
  {
    text: __('Created date'),
    value: 'created',
    sort: {
      asc: 'created_asc',
      desc: 'created_desc',
    },
  },
  {
    text: __('Expiration date'),
    value: 'expires',
    sort: {
      asc: 'expires_asc',
      desc: 'expires_desc',
    },
  },
  {
    text: __('Last used date'),
    value: 'last_used',
    sort: {
      asc: 'last_used_asc',
      desc: 'last_used_desc',
    },
  },
  {
    text: __('Name'),
    value: 'name',
    sort: {
      asc: 'name_asc',
      desc: 'name_desc',
    },
  },
];

export const STATISTICS_CONFIG = [
  {
    title: s__('AccessTokens|Active tokens'),
    tooltipTitle: s__('AccessTokens|Filter for active tokens'),
    filters: [
      {
        type: 'state',
        value: {
          data: 'active',
          operator: '=',
        },
      },
    ],
  },
  {
    title: s__('AccessTokens|Tokens expiring in 2 weeks'),
    tooltipTitle: s__('AccessTokens|Filter for tokens expiring in 2 weeks'),
    filters: [
      {
        type: 'state',
        value: {
          data: 'active',
          operator: '=',
        },
      },
      {
        type: 'expires',
        value: {
          data: 'DATE_HOLDER',
          operator: '<',
        },
      },
    ],
  },
  {
    title: s__('AccessTokens|Revoked tokens'),
    tooltipTitle: s__('AccessTokens|Filter for revoked tokens'),
    filters: [
      {
        type: 'revoked',
        value: {
          data: 'true',
          operator: '=',
        },
      },
    ],
  },
  {
    title: s__('AccessTokens|Expired tokens'),
    tooltipTitle: s__('AccessTokens|Filter for expired tokens'),
    filters: [
      {
        type: 'revoked',
        value: {
          data: 'false',
          operator: '=',
        },
      },
      {
        type: 'state',
        value: {
          data: 'inactive',
          operator: '=',
        },
      },
    ],
  },
];
