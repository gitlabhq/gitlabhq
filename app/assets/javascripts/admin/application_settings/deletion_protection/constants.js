import { __, s__ } from '~/locale';

export const I18N_DELETION_PROTECTION = {
  label: s__('DeletionSettings|Retention period'),
  helpText: s__(
    'DeletionSettings|Period that deleted groups and projects will remain restorable for.',
  ),
  learnMore: __('Learn more.'),
  days: __('days'),
};

export const DEL_ADJ_PERIOD_MAX_LIMIT = 90;
export const DEL_ADJ_PERIOD_MIN_LIMIT = 1;

export const DEL_ADJ_PERIOD_MAX_LIMIT_ERROR = s__(
  'DeletionSettings|Maximum deletion protection duration is 90 days.',
);

export const DEL_ADJ_PERIOD_MIN_LIMIT_ERROR = s__(
  'DeletionSettings|Minimum deletion protection duration is 1 day.',
);
