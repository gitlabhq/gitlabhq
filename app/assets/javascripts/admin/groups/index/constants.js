import groupsEmptyStateIllustration from '@gitlab/svgs/dist/illustrations/empty-state/empty-groups-md.svg?url';
import { s__, __ } from '~/locale';
import {
  SORT_LABEL_NAME,
  SORT_LABEL_CREATED,
  SORT_LABEL_UPDATED,
  SORT_LABEL_STORAGE_SIZE,
} from '~/groups_projects/constants';
import GroupsList from '~/vue_shared/components/groups_list/groups_list.vue';
import { formatGraphQLGroups } from '~/vue_shared/components/groups_list/formatter';
import ResourceListsEmptyState from '~/vue_shared/components/resource_lists/empty_state.vue';
import adminGroupsQuery from './graphql/queries/groups.query.graphql';

const baseTab = {
  formatter: (groups) =>
    formatGraphQLGroups(groups, (group) => {
      const adminPath = `/admin/groups/${group.fullPath}`;

      return {
        avatarLabelLink: adminPath,
        editPath: `${adminPath}/edit`,
      };
    }),
  listComponent: GroupsList,
  listComponentProps: {
    listItemClass: 'gl-px-5',
    showGroupIcon: true,
  },
  emptyStateComponent: ResourceListsEmptyState,
  query: adminGroupsQuery,
  queryPath: 'groups',
};

export const ACTIVE_TAB = {
  ...baseTab,
  text: __('Active'),
  value: 'active',
  variables: { active: true },
  countsQueryPath: 'active',
  emptyStateComponentProps: {
    svgPath: groupsEmptyStateIllustration,
    title: s__("Groups|You don't have any active groups yet."),
    description: s__(
      'Organization|A group is a collection of several projects. If you organize your projects under a group, it works like a folder.',
    ),
    'data-testid': 'groups-empty-state',
  },
};

export const INACTIVE_TAB = {
  ...baseTab,
  text: __('Inactive'),
  value: 'inactive',
  variables: { active: false },
  countsQueryPath: 'inactive',
  emptyStateComponentProps: {
    svgPath: groupsEmptyStateIllustration,
    title: s__("Groups|You don't have any inactive groups."),
    description: s__('Groups|Groups that are archived or pending deletion will appear here.'),
  },
};

export const SORT_OPTION_NAME = {
  value: 'name',
  text: SORT_LABEL_NAME,
};

export const SORT_OPTION_CREATED = {
  value: 'created_at',
  text: SORT_LABEL_CREATED,
};

export const SORT_OPTION_UPDATED = {
  value: 'updated_at',
  text: SORT_LABEL_UPDATED,
};

export const SORT_OPTION_STORAGE_SIZE = {
  value: 'storage_size',
  text: SORT_LABEL_STORAGE_SIZE,
};

export const SORT_OPTIONS = [
  SORT_OPTION_NAME,
  SORT_OPTION_CREATED,
  SORT_OPTION_UPDATED,
  SORT_OPTION_STORAGE_SIZE,
];

export const ADMIN_GROUPS_TABS = [ACTIVE_TAB, INACTIVE_TAB];

export const BASE_ROUTE = '/admin/groups';

export const ADMIN_GROUPS_ROUTE_NAME = 'admin-groups';

export const FIRST_TAB_ROUTE_NAMES = [ADMIN_GROUPS_ROUTE_NAME];

export const FILTERED_SEARCH_TERM_KEY = 'search';
export const FILTERED_SEARCH_NAMESPACE = 'admin-groups';
