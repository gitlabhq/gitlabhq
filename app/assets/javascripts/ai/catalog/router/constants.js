export const AI_CATALOG_INDEX_ROUTE = 'ai-catalog';
