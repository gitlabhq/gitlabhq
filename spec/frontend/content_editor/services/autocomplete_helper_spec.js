import MockAdapter from 'axios-mock-adapter';
import axios from '~/lib/utils/axios_utils';
import AutocompleteHelper, {
  defaultSorter,
  customSorter,
  createDataSource,
} from '~/content_editor/services/autocomplete_helper';
import { HTTP_STATUS_OK, HTTP_STATUS_INTERNAL_SERVER_ERROR } from '~/lib/utils/http_status';
import { EMOJI_THUMBS_UP } from '~/emoji/constants';
import {
  MOCK_MEMBERS,
  MOCK_COMMANDS,
  MOCK_EPICS,
  MOCK_ISSUES,
  MOCK_LABELS,
  MOCK_MILESTONES,
  MOCK_ITERATIONS,
  MOCK_SNIPPETS,
  MOCK_VULNERABILITIES,
  MOCK_MERGE_REQUESTS,
  MOCK_ASSIGNEES,
  MOCK_REVIEWERS,
  MOCK_WIKIS,
  MOCK_NEW_MEMBERS,
} from './autocomplete_mock_data';

jest.mock('~/emoji', () => ({
  initEmojiMap: () => jest.fn(),
  getAllEmoji: () => [{ name: 'thumbsup' }],
}));

jest.mock('~/graphql_shared/issuable_client', () => ({
  currentAssignees: jest.fn().mockReturnValue({
    1: [
      {
        type: 'User',
        username: 'errol',
        name: "Linnie O'Connell",
        avatar_url:
          'https://www.gravatar.com/avatar/d3d9a468a9884eb217fad5ca5b2b9bd7?s=80\u0026d=identicon',
        availability: null,
      },
      {
        type: 'User',
        username: 'evelynn_olson',
        name: 'Dimple Dare',
        avatar_url:
          'https://www.gravatar.com/avatar/bc1e51ee3512c2b4442f51732d655107?s=80\u0026d=identicon',
        availability: null,
      },
    ],
  }),
  linkedItems: jest.fn().mockReturnValue({
    'gitlab-org/gitlab-test:1': [
      {
        iid: 31,
        title: 'rdfhdfj',
        id: null,
        workItemType: {
          __typename: 'WorkItemType',
          id: 'gid://gitlab/WorkItems::Type/1',
          name: 'Issue',
          iconName: 'issue-type-issue',
        },
      },
      {
        iid: 30,
        title: 'incident1',
        id: null,
        workItemType: {
          __typename: 'WorkItemType',
          id: 'gid://gitlab/WorkItems::Type/1',
          name: 'Issue',
          iconName: 'issue-type-issue',
        },
      },
    ],
  }),
}));

describe('defaultSorter', () => {
  it('returns items as is if query is empty', () => {
    const items = [{ name: 'abc' }, { name: 'bcd' }, { name: 'cde' }];
    const sorter = defaultSorter(['name']);
    expect(sorter(items, '')).toEqual(items);
  });

  it('sorts items based on query match', () => {
    const items = [{ name: 'abc' }, { name: 'bcd' }, { name: 'cde' }];
    const sorter = defaultSorter(['name']);
    expect(sorter(items, 'b')).toEqual([{ name: 'bcd' }, { name: 'abc' }, { name: 'cde' }]);
  });

  it('sorts items based on query match in multiple fields', () => {
    const items = [
      { name: 'wabc', description: 'xyz' },
      { name: 'bcd', description: 'wxy' },
      { name: 'cde', description: 'vwx' },
    ];
    const sorter = defaultSorter(['name', 'description']);
    expect(sorter(items, 'w')).toEqual([
      { name: 'wabc', description: 'xyz' },
      { name: 'bcd', description: 'wxy' },
      { name: 'cde', description: 'vwx' },
    ]);
  });
});

describe('customSorter', () => {
  it('sorts items based on custom sorter function', () => {
    const items = [3, 1, 2];
    const sorter = customSorter((a, b) => a - b);
    expect(sorter(items)).toEqual([1, 2, 3]);
  });
});

describe('createDataSource', () => {
  let mock;

  beforeEach(() => {
    mock = new MockAdapter(axios);
  });

  afterEach(() => {
    mock.restore();
  });

  describe('on fetch success', () => {
    const dataSourceParams = {
      source: '/source',
      searchFields: ['name', 'description'],
    };

    beforeEach(() => {
      const data = [
        { name: 'abc', description: 'xyz' },
        { name: 'bcd', description: 'wxy' },
        { name: 'cde', description: 'vwx' },
      ];
      mock.onGet('/source').reply(HTTP_STATUS_OK, data);
    });

    it('fetches data from source and filters based on query', async () => {
      const dataSource = createDataSource(dataSourceParams);

      const results = await dataSource.search('b');
      expect(results).toEqual([
        { name: 'bcd', description: 'wxy' },
        { name: 'abc', description: 'xyz' },
      ]);
    });
  });

  it('handles source fetch errors', async () => {
    mock.onGet('/source').reply(HTTP_STATUS_INTERNAL_SERVER_ERROR);

    const dataSource = createDataSource({
      source: '/source',
      searchFields: ['name', 'description'],
      sorter: (items) => items,
    });

    const results = await dataSource.search('b');
    expect(results).toEqual([]);
  });
});

describe('AutocompleteHelper', () => {
  let mock;
  let autocompleteHelper;
  let dateNowOld;
  let dataSourceUrls;

  beforeEach(() => {
    mock = new MockAdapter(axios);
    dataSourceUrls = {
      members: '/members',
      issues: '/issues',
      snippets: '/snippets',
      labels: '/labels',
      epics: '/epics',
      milestones: '/milestones',
      iterations: '/iterations',
      mergeRequests: '/mergeRequests',
      vulnerabilities: '/vulnerabilities',
      commands: '/commands',
      wikis: '/wikis',
    };

    mock.onGet('/members').reply(HTTP_STATUS_OK, MOCK_MEMBERS);
    mock.onGet('/issues').reply(HTTP_STATUS_OK, MOCK_ISSUES);
    mock.onGet('/snippets').reply(HTTP_STATUS_OK, MOCK_SNIPPETS);
    mock.onGet('/labels').reply(HTTP_STATUS_OK, MOCK_LABELS);
    mock.onGet('/epics').reply(HTTP_STATUS_OK, MOCK_EPICS);
    mock.onGet('/milestones').reply(HTTP_STATUS_OK, MOCK_MILESTONES);
    mock.onGet('/iterations').reply(HTTP_STATUS_OK, MOCK_ITERATIONS);
    mock.onGet('/mergeRequests').reply(HTTP_STATUS_OK, MOCK_MERGE_REQUESTS);
    mock.onGet('/vulnerabilities').reply(HTTP_STATUS_OK, MOCK_VULNERABILITIES);
    mock.onGet('/commands').reply(HTTP_STATUS_OK, MOCK_COMMANDS);
    mock.onGet('/wikis').reply(HTTP_STATUS_OK, MOCK_WIKIS);

    mock.onGet('/new/members').reply(HTTP_STATUS_OK, MOCK_NEW_MEMBERS);

    const sidebarMediator = {
      store: {
        assignees: MOCK_ASSIGNEES,
        reviewers: MOCK_REVIEWERS,
      },
    };

    autocompleteHelper = new AutocompleteHelper({
      dataSourceUrls,
      sidebarMediator,
    });

    dateNowOld = Date.now();

    jest.spyOn(Date, 'now').mockImplementation(() => new Date('2023-11-14').getTime());
  });

  afterEach(() => {
    mock.restore();

    delete gl.GfmAutoComplete;

    jest.spyOn(Date, 'now').mockImplementation(() => dateNowOld);
  });

  it.each`
    referenceType      | query
    ${'user'}          | ${'r'}
    ${'issue'}         | ${'q'}
    ${'snippet'}       | ${'s'}
    ${'label'}         | ${'c'}
    ${'epic'}          | ${'n'}
    ${'milestone'}     | ${'16'}
    ${'iteration'}     | ${'27'}
    ${'merge_request'} | ${'n'}
    ${'vulnerability'} | ${'cross'}
    ${'command'}       | ${'re'}
    ${'wiki'}          | ${'ho'}
  `(
    'for reference type "$referenceType", searches for "$query" correctly',
    async ({ referenceType, query }) => {
      const dataSource = autocompleteHelper.getDataSource(referenceType);
      const results = await dataSource.search(query);

      expect(
        results.map(({ title, name, username }) => username || name || title),
      ).toMatchSnapshot();
    },
  );

  it.each`
    referenceType | command
    ${'label'}    | ${'/label'}
    ${'label'}    | ${'/unlabel'}
    ${'label'}    | ${'/relabel'}
    ${'user'}     | ${'/assign'}
    ${'user'}     | ${'/reassign'}
    ${'user'}     | ${'/unassign'}
    ${'user'}     | ${'/assign_reviewer'}
    ${'user'}     | ${'/unassign_reviewer'}
    ${'user'}     | ${'/reassign_reviewer'}
  `(
    'filters items based on command "$command" for reference type "$referenceType" and command',
    async ({ referenceType, command }) => {
      const dataSource = autocompleteHelper.getDataSource(referenceType, { command });
      const results = await dataSource.search();

      expect(
        results.map(({ username, name, title }) => username || name || title),
      ).toMatchSnapshot();
    },
  );

  describe('for work items', () => {
    beforeEach(() => {
      autocompleteHelper = new AutocompleteHelper({
        dataSourceUrls,
        sidebarMediator: {
          store: { assignees: [], reviewers: [] },
        },
      });

      autocompleteHelper.tiptapEditor = {
        view: {
          dom: {
            closest: () => ({
              dataset: {
                workItemFullPath: 'gitlab-org/gitlab-test',
                workItemId: 1,
                workItemIid: 1,
              },
            }),
          },
        },
      };
    });

    it.each`
      command
      ${'/assign'}
      ${'/unassign'}
    `('filters users using apollo cache for command "$command"', async ({ command }) => {
      const dataSource = autocompleteHelper.getDataSource('user', { command });
      const results = await dataSource.search();

      expect(results.map(({ username }) => username)).toMatchSnapshot();
    });

    it.each`
      command
      ${'/unlink'}
    `('filters work items using apollo cache for command "$command"', async ({ command }) => {
      const dataSource = autocompleteHelper.getDataSource('issue', { command });
      const results = await dataSource.search();

      expect(results.map(({ iid }) => iid)).toMatchSnapshot();
    });
  });

  it('filters items correctly for the second time, when the first command was different', async () => {
    let dataSource = autocompleteHelper.getDataSource('label', { command: '/label' });
    let results = await dataSource.search();

    // all labels listed for the first command
    expect(results.map(({ title }) => title)).toEqual([
      'Bronce',
      'Contour',
      'Corolla',
      'Cygsync',
      'Frontier',
      'Grand Am',
      'Onesync',
      'Phone',
      'Pynefunc',
      'Trinix',
      'Trounswood',
      'group::knowledge',
      'scoped label',
      'type::one',
      'type::two',
    ]);

    dataSource = autocompleteHelper.getDataSource('label', { command: '/unlabel' });
    results = await dataSource.search();

    // only set labels listed for the second command
    expect(results.map(({ title }) => title)).toEqual(['Amsche', 'Brioffe', 'Bryncefunc', 'Ghost']);
  });

  it('loads default datasources if not passed', () => {
    gl.GfmAutoComplete = {
      dataSources: {
        members: '/gitlab-org/gitlab-test/-/autocomplete_sources/members',
      },
    };
    autocompleteHelper = new AutocompleteHelper({});

    expect(autocompleteHelper.dataSourceUrls.members).toBe(
      '/gitlab-org/gitlab-test/-/autocomplete_sources/members',
    );
  });

  it("loads emoji if dataSources doesn't exist", async () => {
    autocompleteHelper = new AutocompleteHelper({});

    const dataSource = autocompleteHelper.getDataSource('emoji');
    const results = await dataSource.search('');

    expect(results).toEqual([{ emoji: { name: EMOJI_THUMBS_UP }, fieldValue: EMOJI_THUMBS_UP }]);
  });

  it('updates dataSourcesUrl correctly', () => {
    const newDataSources = {
      members: '/new/members',
      issues: '/new/issues',
      snippets: '/new/snippets',
      labels: '/new/labels',
      epics: '/new/epics',
      milestones: '/new/milestones',
      iterations: '/new/iterations',
      mergeRequests: '/new/mergeRequests',
      vulnerabilities: '/new/vulnerabilities',
      commands: '/new/commands',
      wikis: '/new/wikis',
    };

    autocompleteHelper.updateDataSources(newDataSources);
    expect(autocompleteHelper.dataSourceUrls).toEqual(newDataSources);
  });

  it('returns expected results before and after updating data sources', async () => {
    // Retrieve the initial data source and search for 'user'
    let dataSource = autocompleteHelper.getDataSource('user');
    let results = await dataSource.search('');

    expect(results.map(({ username }) => username)).toMatchSnapshot();

    // Update the data sources
    const newDataSources = {
      members: '/new/members',
    };
    autocompleteHelper.updateDataSources(newDataSources);

    // Retrieve the updated data source and search for 'user'
    dataSource = autocompleteHelper.getDataSource('user');
    results = await dataSource.search('');

    expect(results.map(({ username }) => username)).toMatchSnapshot();
  });

  describe('if filterOnBackend: true', () => {
    it('fetches data passing a `search` param', async () => {
      const dataSource = autocompleteHelper.getDataSource('user', { filterOnBackend: true });

      await dataSource.search('bcd');

      expect(mock.history.get[0].params).toEqual({ search: 'bcd' });
    });
  });
});
