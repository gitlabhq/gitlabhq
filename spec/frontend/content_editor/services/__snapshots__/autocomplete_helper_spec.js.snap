// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AutocompleteHelper filters items based on command "/assign" for reference type "user" and command 1`] = `
[
  "florida.schoen",
  "root",
  "all",
  "lakeesha.batz",
  "laurene_blick",
  "myrtis",
  "patty",
  "Commit451",
  "flightjs",
  "gitlab-instance-ade037f9",
  "gitlab-org",
  "gnuwget",
  "h5bp",
  "jashkenas",
  "twitter",
]
`;

exports[`AutocompleteHelper filters items based on command "/assign_reviewer" for reference type "user" and command 1`] = `
[
  "florida.schoen",
  "root",
  "all",
  "errol",
  "evelynn_olson",
  "Commit451",
  "flightjs",
  "gitlab-instance-ade037f9",
  "gitlab-org",
  "gnuwget",
  "h5bp",
  "jashkenas",
  "twitter",
]
`;

exports[`AutocompleteHelper filters items based on command "/label" for reference type "label" and command 1`] = `
[
  "Bronce",
  "Contour",
  "Corolla",
  "Cygsync",
  "Frontier",
  "Grand Am",
  "Onesync",
  "Phone",
  "Pynefunc",
  "Trinix",
  "Trounswood",
  "group::knowledge",
  "scoped label",
  "type::one",
  "type::two",
]
`;

exports[`AutocompleteHelper filters items based on command "/reassign" for reference type "user" and command 1`] = `
[
  "florida.schoen",
  "root",
  "all",
  "errol",
  "evelynn_olson",
  "lakeesha.batz",
  "laurene_blick",
  "myrtis",
  "patty",
  "Commit451",
  "flightjs",
  "gitlab-instance-ade037f9",
  "gitlab-org",
  "gnuwget",
  "h5bp",
]
`;

exports[`AutocompleteHelper filters items based on command "/reassign_reviewer" for reference type "user" and command 1`] = `
[
  "florida.schoen",
  "root",
  "all",
  "errol",
  "evelynn_olson",
  "lakeesha.batz",
  "laurene_blick",
  "myrtis",
  "patty",
  "Commit451",
  "flightjs",
  "gitlab-instance-ade037f9",
  "gitlab-org",
  "gnuwget",
  "h5bp",
]
`;

exports[`AutocompleteHelper filters items based on command "/relabel" for reference type "label" and command 1`] = `
[
  "Amsche",
  "Brioffe",
  "Bronce",
  "Bryncefunc",
  "Contour",
  "Corolla",
  "Cygsync",
  "Frontier",
  "Ghost",
  "Grand Am",
  "Onesync",
  "Phone",
  "Pynefunc",
  "Trinix",
  "Trounswood",
]
`;

exports[`AutocompleteHelper filters items based on command "/unassign" for reference type "user" and command 1`] = `
[
  "errol",
  "evelynn_olson",
]
`;

exports[`AutocompleteHelper filters items based on command "/unassign_reviewer" for reference type "user" and command 1`] = `
[
  "lakeesha.batz",
  "laurene_blick",
  "myrtis",
  "patty",
]
`;

exports[`AutocompleteHelper filters items based on command "/unlabel" for reference type "label" and command 1`] = `
[
  "Amsche",
  "Brioffe",
  "Bryncefunc",
  "Ghost",
]
`;

exports[`AutocompleteHelper for reference type "command", searches for "re" correctly 1`] = `
[
  "relabel",
  "remove_milestone",
  "remove_estimate",
  "remove_time_spent",
  "relate",
  "remove_epic",
  "reassign",
  "create_merge_request",
]
`;

exports[`AutocompleteHelper for reference type "epic", searches for "n" correctly 1`] = `
[
  "Nobis quidem aspernatur reprehenderit sunt ut ipsum tempora sapiente sed iste.",
  "Minus eius ut omnis quos sunt dicta ex ipsum.",
  "Quae nostrum possimus rerum aliquam pariatur a eos aut id.",
  "Dicta incidunt vel dignissimos sint sit esse est quibusdam quidem consequatur.",
  "Doloremque a quisquam qui culpa numquam doloribus similique iure enim.",
]
`;

exports[`AutocompleteHelper for reference type "issue", searches for "q" correctly 1`] = `
[
  "Quasi id et et nihil sint autem.",
  "Eaque omnis eius quas necessitatibus hic ut et corrupti.",
  "Aut quisquam magnam eos distinctio incidunt perferendis fugit.",
  "Dolorem quisquam cupiditate consequatur perspiciatis sequi eligendi ullam.",
  "Nesciunt quia molestiae in aliquam amet et dolorem.",
  "Porro tempore qui qui culpa saepe et nam quos.",
  "Sed sint a est consequatur quae quasi autem debitis alias.",
  "Molestiae minima maxime optio nihil quam eveniet dolor.",
  "Et laboriosam aut ratione voluptatem quasi recusandae.",
  "Et molestiae delectus voluptates velit vero illo aut rerum quo et.",
]
`;

exports[`AutocompleteHelper for reference type "iteration", searches for "27" correctly 1`] = `
[
  "Optio quod at quia ad pariatur dolores. Nov 26, 2024 - Dec 9, 2024",
  "Itaque repellat possimus dolor quia nesciunt hic aut iure animi vel. Nov 13, 2024 - Dec 10, 2024",
  "Magni eos minus vero facilis in consequuntur deserunt omnis consectetur natus. Nov 27, 2024 - Dec 10, 2024",
]
`;

exports[`AutocompleteHelper for reference type "label", searches for "c" correctly 1`] = `
[
  "Contour",
  "Corolla",
  "Cygsync",
  "scoped label",
  "Amsche",
  "Bronce",
  "Bryncefunc",
  "Onesync",
  "Pynefunc",
]
`;

exports[`AutocompleteHelper for reference type "merge_request", searches for "n" correctly 1`] = `
[
  "Blanditiis maxime voluptatem ut pariatur vel autem vero non quod libero.",
  "Optio nemo qui dolorem sit ipsum qui saepe.",
  "Draft: Alunny/publish lib",
  "Draft: Fix event current target",
  "Draft: Resolve "hgvbbvnnb"",
  "Autem eaque et sed provident enim corrupti molestiae.",
  "Always call registry's trigger method from withRegistration",
]
`;

exports[`AutocompleteHelper for reference type "milestone", searches for "16" correctly 1`] = `
[
  "16.7",
  "16.8",
  "16.9",
  "16.10",
  "16.11",
  "16.0",
  "16.1",
  "16.2",
  "16.3",
  "16.4",
  "16.5",
  "16.6",
]
`;

exports[`AutocompleteHelper for reference type "snippet", searches for "s" correctly 1`] = `
[
  "ss",
  "test snippet",
  "another test snippet",
]
`;

exports[`AutocompleteHelper for reference type "user", searches for "r" correctly 1`] = `
[
  "root",
  "errol",
  "lakeesha.batz",
  "myrtis",
  "florida.schoen",
  "laurene_blick",
  "all",
  "twitter",
  "gitlab-org",
  "evelynn_olson",
]
`;

exports[`AutocompleteHelper for reference type "vulnerability", searches for "cross" correctly 1`] = `
[
  "Cross Site Scripting (Persistent)",
  "Cross Site Scripting (Persistent)",
  "Cross Site Scripting (Persistent)",
]
`;

exports[`AutocompleteHelper for reference type "wiki", searches for "ho" correctly 1`] = `
[
  "Home",
  "How to use GitLab",
]
`;

exports[`AutocompleteHelper for work items filters users using apollo cache for command "/assign" 1`] = `
[
  "florida.schoen",
  "root",
  "all",
  "lakeesha.batz",
  "laurene_blick",
  "myrtis",
  "patty",
  "Commit451",
  "flightjs",
  "gitlab-instance-ade037f9",
  "gitlab-org",
  "gnuwget",
  "h5bp",
  "jashkenas",
  "twitter",
]
`;

exports[`AutocompleteHelper for work items filters users using apollo cache for command "/unassign" 1`] = `
[
  "errol",
  "evelynn_olson",
]
`;

exports[`AutocompleteHelper for work items filters work items using apollo cache for command "/unlink" 1`] = `
[
  31,
  30,
]
`;

exports[`AutocompleteHelper returns expected results before and after updating data sources 1`] = `
[
  "florida.schoen",
  "root",
  "all",
  "errol",
  "evelynn_olson",
  "lakeesha.batz",
  "laurene_blick",
  "myrtis",
  "patty",
  "Commit451",
  "flightjs",
  "gitlab-instance-ade037f9",
  "gitlab-org",
  "gnuwget",
  "h5bp",
]
`;

exports[`AutocompleteHelper returns expected results before and after updating data sources 2`] = `
[
  "walker.finn",
]
`;
